model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java

```diff
@@ -0,0 +1,139 @@
+
+
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.common.DeepCloner;
+import org.keycloak.models.map.common.StringKeyConverter;
+import org.keycloak.models.map.common.StringKeyConverter.StringKey;
+import org.keycloak.models.map.common.UpdatableEntity;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.ModelEntityUtil;
+import org.keycloak.models.map.storage.chm.ConcurrentHashMapKeycloakTransaction;
+import org.keycloak.models.map.storage.chm.MapFieldPredicates;
+import org.keycloak.models.map.storage.chm.MapModelCriteriaBuilder.UpdatePredicatesFunc;
+import org.keycloak.storage.SearchableModelField;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * {@link MapKeycloakTransaction} implementation used with the file map storage.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class FileMapKeycloakTransaction<V extends AbstractEntity & UpdatableEntity, M>
+  extends ConcurrentHashMapKeycloakTransaction<String, V, M> {
+
+    private final List<Path> touchedPaths = new LinkedList<>();
+    private Map<Path, Path> renameOnCommit = new IdentityHashMap<>();
+
+    private final String txId = StringKey.INSTANCE.yieldNewUniqueKey();
+
+    public static <V extends AbstractEntity & UpdatableEntity, M> FileMapKeycloakTransaction<V, M> newInstance(Class<V> entityClass,
+      Function<String, Path> dataDirectoryFunc, Function<V, String[]> suggestedPath,
+      boolean isExpirableEntity, Map<SearchableModelField<? super M>, UpdatePredicatesFunc<String, V, M>> fieldPredicates) {
+        Crud<V, M> crud = new Crud<>(entityClass, dataDirectoryFunc, suggestedPath, isExpirableEntity, fieldPredicates);
+        FileMapKeycloakTransaction<V, M> tx = new FileMapKeycloakTransaction<>(entityClass, crud);
+        crud.tx = tx;
+        return tx;
+    }
+
+    private FileMapKeycloakTransaction(Class<V> entityClass, Crud crud) {
+        super(
+          crud,
+          StringKeyConverter.StringKey.INSTANCE,
+          DeepCloner.DUMB_CLONER,
+          MapFieldPredicates.getPredicates(ModelEntityUtil.getModelType(entityClass)),
+          ModelEntityUtil.getRealmIdField(entityClass)
+        );
+    }
+
+    @Override
+    public void rollback() {
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+        super.rollback();
+    }
+
+    @Override
+    public void commit() {
+        super.commit();
+        renameOnCommit.forEach(FileMapKeycloakTransaction::silentMove);
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+    }
+
+    private static void silentMove(Path from, Path to) {
```

Comment:
Those `silent()` methods swallow exceptions, which can be dangerous. I don't see any JavaDoc indicating when it should be used and when not. Please advise when it is a good idea to swallow those. I just assume bad things can happend when something isn't deleted that is supposed to be deleted, or not moved when it was supposed to be moved.

Reply:
These are private methods of `FileMapKeycloakTransaction`. I've updated the `silentMove` to throw an exception if it cannot finish its job as it can spoil the filesystem, but failed removal does not lead to inconsistency.